### PR TITLE
chore(platforms): Migrate  x86 gnu target to cross, add new aarch64 gnu target (no sasl)

### LINKED ISSUE #3638: Transition the `x86_64-unknown-linux-gnu` target to the new release process using cross
As a first step towards resolving our releasing woes, we want to lay a solid foundation by supporting the `x86_64-unknown-linux-gnu` **only**. This process should serve as an example for migrating other targets. It should set boundaries, constraints, and demonstrate what a sane release process looks like. Once this is set, we will follow up with supporting other architectures.

Cross was noted in the release title since this was discussed as the proper foundation going forward.

### LINKED ISSUE #3811: Transition the `aarch64-unknown-linux-gnu` target to the new release process using cross
Building off #3638, enable aarch64 gnu as a precursor to enabling aarch64 musl.

### PR DESCRIPTION
**Pull request by Hoverbear**
Builds a bit alongside the ideas of  #3537.

Closes #3638.
Closes #3811. 
Causes #3805.

### CODE REVIEW

**Comment by Hoverbear** [comment]
Related to #3537

File: Makefile [hunk 1]
```diff
@@ -161,22 +167,83 @@ build-dev: ## Build the project in development mode (Supports `ENVIRONMENT=true`
...
+##@ Cross Compiling
+.PHONY: cross-enable
+cross-enable: cargo-install-cross
+
+.PHONY: CARGO_HANDLES_FRESHNESS
+CARGO_HANDLES_FRESHNESS:
+	${EMPTY}
+
+# This is basically a shorthand for folks.
+# `cross-anything-triple` will call `cross anything --target triple` with the right features.
```
Comment (jszwedko) [inline Makefile:191]: :+1:

File: Makefile [hunk 2]
```diff
@@ -161,22 +167,83 @@ build-dev: ## Build the project in development mode (Supports `ENVIRONMENT=true`
...
+cross-%: export PROFILE ?= release
+cross-%: export RUSTFLAGS += -C link-arg=-s
+cross-%: cargo-install-cross
+	cross ${COMMAND} \
+		$(if $(findstring release,$(PROFILE)),--release,) \
+		--target ${TRIPLE} \
+		--no-default-features \
+		--features target-${TRIPLE}
+
+target/%/vector: export PAIR =$(subst /, ,$(@:target/%/vector=%))
```
Comment (jszwedko) [inline Makefile:205]: TIL you can set target specific variables like this :clap:

**Comment by jszwedko** [comment]
Noting that from the call we expect glibc to be dynamically linked. It isn't clear if the rest of the above ones fall in that bucket or not.

These are the linked libraries I see for aarch64:

```
 /usr/aarch64-linux-gnu/bin/readelf -d target/aarch64-unknown-linux-gnu/release/vector | grep 'Shared library'
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-aarch64.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
```

Are these the expected ones?

**Comment by Hoverbear** [comment]
@jszwedko I'm noting this:

image

Hm, I wonder how much we value centos:7 support?

**Comment by Hoverbear** [comment]
@jszwedko This seems expected:
image

**Comment by jszwedko** [comment]
> @jszwedko I'm noting this:
> 
> image
> Hm, I wonder how much we value centos:7 support?

Hmm. I feel like it'd be nice to given that `centos:7` is supported until 2024. I tested out:

* `ubuntu:14.04`
* `fedora:31`
* `debian:9`

All of which seemed fine. This seems like something we might want to have a build matrix to verify supported OSes for.

**Comment by Hoverbear** [comment]
No problem, I think we have this https://github.com/timberio/vector/blob/cross-env/scripts/environment/bootstrap/centos-7.sh I can use.

**Comment by Hoverbear** [comment]
We decided not to use a custom image since we identified https://github.com/rust-embedded/cross/issues/455

**Comment by Hoverbear** [comment]
I'm going to explore re-enabling sasl a bit before I prep a musl PR.

File: Makefile (hunk 2 above)
Comment (Hoverbear) [reply Makefile:205]: You can also do `?=` to only set if not set, and `override` to force a specific variable to be set. One note though, the envs you export here you can't use in the target dependencies below. :(

**Comment by Hoverbear** [comment]
No dice. I'm going to base a new musl PR off this, but this is ready for review. :)

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -71,6 +72,44 @@ jobs:
...
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+      - run: sudo bash scripts/environment/bootstrap-ubuntu-20.04.sh
+      - run: bash scripts/environment/prepare.sh
+      - run: make slim-builds
+      - run: make build-x86_64-unknown-linux-gnu
```
Comment (jszwedko) [inline .github/workflows/tests.yml:111]: This should be `make build-aarch64-unknown-linux-gnu`, no?

**Review by jszwedko — APPROVED** [review]
Great work! I like how this is coming together.

Just one note about the CI target for the `build-aarch64-unknown-linux-gnu` action that should be resolved, otherwise :+1:

**Comment by Hoverbear** [comment]
@jszwedko The final feedback is reflected. :) I'm going to move onto the musl builds and try to get them going.

**Review by jamtur01 — APPROVED** [review]
LGTM.

**Comment by Hoverbear** [comment]
Note this disables sasl2 support on GNU AARCH builds for now. I'd still like to fix this.

**Comment by Hoverbear** [comment]
Trying with `rdkafka-plain`:

```
error: linking with `aarch64-linux-gnu-gcc` failed: exit code: 1
  |
  = note: "aarch64-linux-gnu-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/rust/lib/rustlib/aarch64-unknown-linux-gnu/lib" "/target/aarch64-unknown-linux-gnu/release/deps/vector-db28a900f7c911d5.vector.4u6vtfy7-cgu.0.rcgu.o" "-o" "/target/aarch64-unknown-linux-gnu/release/deps/vector-db28a900f7c911d5" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/target/aarch64-unknown-linux-gnu/release/deps" "-L" "/target/release/deps" "-L" "/target/aarch64-unknown-linux-gnu/release/build/ring-3f4a20ff3d88ddda/out" "-L" "/target/aarch64-unknown-linux-gnu/release/build/openssl-sys-f9b600350e7a6c54/out/openssl-build/install/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/onig_sys-027ad72c317ea270/out" "-L" "/target/aarch64-unknown-linux-gnu/release/build/jemalloc-sys-6d5c40e0184d9cea/out/build/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/leveldb-sys-a8773a1e7c1cdf45/out/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/leveldb-sys-a8773a1e7c1cdf45/out/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/rdkafka-sys-b70a193eb7e7ebd4/out/src" "-L" "/target/aarch64-unknown-linux-gnu/release/build/libz-sys-43160c6dea1507bf/out/build" "-L" "/target/aarch64-unknown-linux-gnu/release/build/sasl2-sys-c95258f429942819/out/install/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/krb5-src-8d1ffec509dacae0/out/install/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/zstd-sys-ac91a4ac0b306564/out" "-L" "/target/aarch64-unknown-linux-gnu/release/build/rlua-862904633f1916ba/out" "-L" "/rust/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/rustcr70Qk2/libjemalloc_sys-4a3385324afff237.rlib" "/tmp/rustcr70Qk2/librlua-7d7dfca4ef2d2e16.rlib" "/tmp/rustcr70Qk2/libonig_sys-99ab780281329b7c.rlib" "/tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib" "/tmp/rustcr70Qk2/libzstd_sys-2c75c3cd2b0eec36.rlib" "/tmp/rustcr70Qk2/liblibz_sys-f0c9cc561d8bb536.rlib" "/tmp/rustcr70Qk2/libsasl2_sys-d8cf86c95a1c830d.rlib" "/tmp/rustcr70Qk2/libopenssl_sys-7715e0c7b7afbf3f.rlib" "/tmp/rustcr70Qk2/libring-bbf8d3ae6c6c2248.rlib" "/tmp/rustcr70Qk2/libleveldb_sys-345abc0782ae8d79.rlib" "-Wl,--start-group" "/tmp/rustcr70Qk2/libbacktrace_sys-9f0c126c25aff795.rlib" "-Wl,--end-group" "/rust/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-df0525f541f6cecf.rlib" "-Wl,-Bdynamic" "-lpthread" "-lresolv" "-lstdc++" "-lutil" "-ldl" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-ldl" "-lutil"
  = note: /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib(rdkafka.o): Relocations in generic ELF (EM: 62)
          /tmp/rustcr70Qk2/librdkafka_sys-69614a9d246dfdfc.rlib: error adding symbols: File in wrong format
          collect2: error: ld returned 1 exit status
          
```

With `rdkafka-cmake`:

```
error: linking with `aarch64-linux-gnu-gcc` failed: exit code: 1
  |
  = note: "aarch64-linux-gnu-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/rust/lib/rustlib/aarch64-unknown-linux-gnu/lib" "/target/aarch64-unknown-linux-gnu/release/deps/vector-bc91ba3c5f6b7d56.vector.2xgcrz2h-cgu.0.rcgu.o" "-o" "/target/aarch64-unknown-linux-gnu/release/deps/vector-bc91ba3c5f6b7d56" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/target/aarch64-unknown-linux-gnu/release/deps" "-L" "/target/release/deps" "-L" "/target/aarch64-unknown-linux-gnu/release/build/ring-3f4a20ff3d88ddda/out" "-L" "/target/aarch64-unknown-linux-gnu/release/build/openssl-sys-f9b600350e7a6c54/out/openssl-build/install/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/onig_sys-027ad72c317ea270/out" "-L" "/target/aarch64-unknown-linux-gnu/release/build/jemalloc-sys-6d5c40e0184d9cea/out/build/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/leveldb-sys-a8773a1e7c1cdf45/out/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/leveldb-sys-a8773a1e7c1cdf45/out/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/rdkafka-sys-0710d5c30109e9d9/out/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/libz-sys-43160c6dea1507bf/out/build" "-L" "/target/aarch64-unknown-linux-gnu/release/build/sasl2-sys-c95258f429942819/out/install/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/krb5-src-8d1ffec509dacae0/out/install/lib" "-L" "/target/aarch64-unknown-linux-gnu/release/build/zstd-sys-ac91a4ac0b306564/out" "-L" "/target/aarch64-unknown-linux-gnu/release/build/rlua-862904633f1916ba/out" "-L" "/rust/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/rustceMcgjI/libjemalloc_sys-4a3385324afff237.rlib" "/tmp/rustceMcgjI/librlua-7d7dfca4ef2d2e16.rlib" "/tmp/rustceMcgjI/libonig_sys-99ab780281329b7c.rlib" "/tmp/rustceMcgjI/librdkafka_sys-e5b074f31e42fef7.rlib" "/tmp/rustceMcgjI/libzstd_sys-2c75c3cd2b0eec36.rlib" "/tmp/rustceMcgjI/liblibz_sys-f0c9cc561d8bb536.rlib" "/tmp/rustceMcgjI/libsasl2_sys-d8cf86c95a1c830d.rlib" "/tmp/rustceMcgjI/libopenssl_sys-7715e0c7b7afbf3f.rlib" "/tmp/rustceMcgjI/libring-bbf8d3ae6c6c2248.rlib" "/tmp/rustceMcgjI/libleveldb_sys-345abc0782ae8d79.rlib" "-Wl,--start-group" "/tmp/rustceMcgjI/libbacktrace_sys-9f0c126c25aff795.rlib" "-Wl,--end-group" "/rust/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-df0525f541f6cecf.rlib" "-Wl,-Bdynamic" "-lpthread" "-lresolv" "-lstdc++" "-lutil" "-ldl" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-ldl" "-lutil"
  = note: /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustceMcgjI/libsasl2_sys-d8cf86c95a1c830d.rlib(client.o): Relocations in generic ELF (EM: 62)
          /usr/lib/gcc-cross/aarch64-linux-gnu/5/../../../../aarch64-linux-gnu/bin/ld: /tmp/rustceMcgjI/libsasl2_sys-d8cf86c95a1c830d.rlib(client.o): Relocations in generic ELF (EM: 62)
          /tmp/rustceMcgjI/libsasl2_sys-d8cf86c95a1c830d.rlib: error adding symbols: File in wrong format
          collect2: error: ld returned 1 exit status
          


```

**Pull request merged**